### ISSUE TITLE
this.property assignment in JS often crashes the language service

### ISSUE THREAD

**Issue by mjbvz** [Microsoft/TypeScript#22774]
_From @pocesar on March 21, 2018 21:20_

Version 1.22.0-insider
Commit 31c30b99345e65fffdc33066726d0962dfbecfa3
Date 2018-03-21T06:17:39.707Z
Shell 1.7.12
Renderer 58.0.3029.110
Node 7.9.0
Architecture x64

Steps to Reproduce:

1. Open an ES7 file
2. Use the following jsconfig.json:

```json
{
  "compilerOptions": {
    "checkJs": true,
    "strict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitAny": false,
    "module": "commonjs",
    "target": "es2017",
    "alwaysStrict": true,
    "moduleResolution": "node",
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "noImplicitUseStrict": true,
    "strictNullChecks": false,
    "baseUrl": "./",
    "rootDir": "./",
    "allowSyntheticDefaultImports": true
  }
}
```  
3. Variables that are defined at the top of the file are reporting as "not being read" (with a wriggly green line below), `require('./middleware')` is a wrong path, when I changed it to '../../middleware', it still shows as being './middleware'

Output > Typescript is returning this error:

```
[Error  - 18:14:55] 'quickinfo' request failed with error.
Error processing request. Cannot read property 'members' of undefined
TypeError: Cannot read property 'members' of undefined
    at bindThisPropertyAssignment (C:\nodejs\node_modules\typescript\lib\tsserver.js:20673:66)
    at bindWorker (C:\nodejs\node_modules\typescript\lib\tsserver.js:20432:29)
    at bind (C:\nodejs\node_modules\typescript\lib\tsserver.js:20336:13)
    at visitNode (C:\nodejs\node_modules\typescript\lib\tsserver.js:13322:24)
    at Object.forEachChild (C:\nodejs\node_modules\typescript\lib\tsserver.js:13547:24)
    at bindEachChild (C:\nodejs\node_modules\typescript\lib\tsserver.js:19294:16)
    at bindChildrenWorker (C:\nodejs\node_modules\typescript\lib\tsserver.js:19383:21)
    at bindChildren (C:\nodejs\node_modules\typescript\lib\tsserver.js:19269:17)
    at bind (C:\nodejs\node_modules\typescript\lib\tsserver.js:20342:21)
    at bindEach (C:\nodejs\node_modules\typescript\lib\tsserver.js:19286:21)
    at visitNodes (C:\nodejs\node_modules\typescript\lib\tsserver.js:13327:24)
    at Object.forEachChild (C:\nodejs\node_modules\typescript\lib\tsserver.js:13536:24)
    at bindEachChild (C:\nodejs\node_modules\typescript\lib\tsserver.js:19294:16)
    at bindChildrenWorker (C:\nodejs\node_modules\typescript\lib\tsserver.js:19383:21)
    at bindChildren (C:\nodejs\node_modules\typescript\lib\tsserver.js:19269:17)
    at bindContainer (C:\nodejs\node_modules\typescript\lib\tsserver.js:19250:17)
    at bind (C:\nodejs\node_modules\typescript\lib\tsserver.js:20345:21)
```

PS: When doing the `Reload Window` it takes forever for the file to start being checked (30s+)

Does this issue occur when all extensions are disabled?: Yes

_Copied from original issue: Microsoft/vscode#46292_

**Comment by mjbvz** [comment]
_From @pocesar on March 21, 2018 21:26_

ok, it seems to be a problem with TS 2.8.0-dev.20180321, it's working as expected with 2.7.2

**Comment by mjbvz** [comment]
@pocesar Can you please share the file that triggers this? May be that a specific syntax structure  causes problems

**Comment by pocesar** [comment]
this one (in `app/controllers/admin/contact.js`):

```js
'use strict'

const express = require('express')
const router = express.Router()
const db = require('../../../config/db')
const q = require('../../../utils/query')
const helpers = require('../../helpers')
const middleware = require('./middleware')

module.exports = function (app) {
  app.use('/admin/contact', middleware, router)
}

router.get('/', async function (_req, res) {
  try {
    const contacts = await db.pgp.task((sco) => (q.getAllContact(sco)))

    res.render('contact/list', {
      title: 'Contatos',
      subtitle: 'Contatos',
      contacts: contacts || []
    })
  } catch (error) {
    helpers.jsonAnswer({
      res,
      error: helpers.BTError.wrapError(
        error,
        400,
        'Ocorreu um erro ao buscar os contatos, tente novamente!'
      )
    })
  }
})
```

**Comment by DanielRosenwasser** [comment]
Do you have a more complete project that reproduces the issue? Something on GitHub perhaps?

**Comment by sandersn** [comment]
I'm working on this. It's partially addressed by #22725. I'll have a second PR out shortly that fixes additional failures that I found in chrome-devtools-frontend and npm's source.

**Comment by sandersn** [comment]
(It was introduced in b9f60566d09 on 2018-03-16 if you need to get back to a working version.)
